design/20230302.gomod.md

```diff
@@ -0,0 +1,244 @@
+
+There's a vulnerability reported for Helm v3.10.3 ([1]) which we'd like to patch, but the only version with a fix
+available is Helm v3.11.1.
+
+Between Helm 3.10 and 3.11, several of Helm's dependencies were upgraded, and crucially Helm has some of the same
+dependencies that cert-manager does. That means that we can't easily _just_ upgrade Helm.
+
+Running `go get -u helm.sh/helm/v3` produces 56 different upgrades of _other_ dependencies. Notably, it bumps our
+Kubernetes dependencies from v0.25.2 to v0.26.0 but there are several other changes.
+
+(NB: Helm is just an example here and we could have problems with any package)
+
+## Proposed Solution: Go Module Proliferation
+
+**In short:** Add several new `go.mod` files so individual components can be patched independently
+
+We can create several new Go modules so that each binary we build can have distinct dependencies. This would mean that
+`cmctl` having a dependency on Helm would only affect `cmctl` and wouldn't force us to change any of the other
+components we build in order to patch a Helm vulnerability.
+
+In addition to separating out the binaries, we could also separate out the "core" issuer implementations into a
+distinct module (since they bring so many external dependencies and currently live in `pkg/` where they'll be part of
+the core cert-manager `go.mod` file).
+
+Plus, where we have testing-only dependencies (e.g. for integration or end-to-end tests) we could create a test module
+so that those test dependencies don't pollute the main `go.mod`.
+
+## Potential Issues
+
+### Importing cert-manager / Development Experience
+
+**In short:** Module replacements help developers but don't aren't respected by imports, which means some changes could need two PRs
+
+**Useful Reference:** It helps to read [this StackOverflow comment](https://stackoverflow.com/a/71984158) to better understand the options we have
+
+The simplest development experience when working with multiple Go modules at once is to use either the `replace` directive
+in `go.mod` `use` in `go.work` to point to local versions of a module. This allows both modules to be developed in
+parallel on a local machine.
+
+In practice, we could look at the `controller`, which would depend on the core cert-manager module. Its `go.mod` could
+contain:
+
+```text
+module github.com/cert-manager/cert-manager/cmd/controller
+
+...
+
+require (
+	github.com/cert-manager/cert-manager v0.0.0-00010101000000-000000000000
+	...
+)
+
+...
+
+replace github.com/cert-manager/cert-manager => ../../
+
+...
+```
+
+This would ensure that the local version of cert-manager is used during development, but that replace wouldn't
+be respected if someone imported `github.com/cert-manager/cert-manager/cmd/controller` from a 3rd party project and
+that 3rd party would see an error relating to the unknown version of cert-manager.
+
+Because of this, changes to `cmd/controller` which also require a change to the core module would likely have to be
```

Comment:
If we have a local replace and the `v0.0.0-00010101000000-000000000000` version, we don't need 2 PRs right?
We only need two PRs if we set the version to eg. `v1.11.0` so the module can be easily imported from external libraries.

Reply:
Yeah that's correct. The more I worked on this the more it felt like breaking imports would be a shame.

I should probably talk more about how replaces would work in practice